src/R/Core/Impl/Tokens/Number.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Globalization;
+
+namespace Microsoft.R.Core.Tokens {
+    public static class Number {
+        private const NumberStyles _hexStyle = NumberStyles.HexNumber | NumberStyles.AllowHexSpecifier;
+        private const NumberStyles _integerStyle = NumberStyles.Integer;
+        private const NumberStyles _doubleStyle = NumberStyles.Number | NumberStyles.AllowExponent;
+
+        public static bool TryParse(string text, out double doubleResult, bool allowLSuffix = true) {
+            doubleResult = Double.NaN;
+            int sign = 1;
+
+            if (text.Length == 0) {
+                return false;
+            }
+            if (allowLSuffix && text[text.Length - 1] == 'L') {
+                text = text.Substring(0, text.Length - 1);
+            }
+
+            if (text.Length == 0) {
+                return false;
+            }
+            if (text[0] == '-' || text[0] == '+') {
```

Comment:
The primary reason it is done this way is colorization which goes by tokens. So -1 gets colorized as a number together with minus. It is possible to always treat it as an operator (except in complex types and exponents) but then if one selects different colors for numbers and operators they will see - as operator which is not what is expected. Although C# doesn't seem to care

![image](https://cloud.githubusercontent.com/assets/12820357/14032742/481c6d28-f1d2-11e5-82e6-750a30bfc4f0.png)

and neither does C++ or JavaScript.